core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java

```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
     Assert.assertTrue("FILE_A should be deleted", deletedFiles.contains(FILE_A.path().toString()));
     Assert.assertTrue("FILE_B should be deleted", deletedFiles.contains(FILE_B.path().toString()));
   }
+
+  /**
+   * Test on table below, and expiring the staged commit `B` using `expireOlderThan` API.
+   * Table: A - C
+   *          ` B (staged)
+   */
+  @Test
+  public void testWithExpiringDanglingStageCommit() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` staged commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    TableMetadata base = readMetadata();
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // `C` commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire all commits including dangling staged snapshot.
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireOlderThan(snapshotB.timestampMillis() + 1)
+        .commit();
+
+    // Files should be deleted of dangling staged snapshot
+    snapshotB.addedFiles().forEach(item -> {
```

Comment:
I think it would be more clear to use a `Set<String>` and validate using `assertEquals`. That way it verifies no additional files were deleted.

```java
Set<String> expectedDeletes = snapshotB.addedFiles().stream()
    .map(added -> added.path().toString())
    .collect(Collectors.toSet());
Assert.assertEquals("Should delete only files added in snapshot B", expectedDeletes, deletedFiles);
```

Reply:
+1, it would be good to add exact files, after considering following.
1. Apart from dataFile, we do delete manifests as well for dangling snapshots
2. The order in which the files are deleted, shouldn't matter, so I have handling that as well.

I think, we should have couple of test to check manifest deletes in case of normal remove snapshots API. Let me know, I can log a ticket explicitly for that, if you agree.